### PR TITLE
Move create document to pipeline architecture

### PR DESCRIPTION
**Pull request by rylev**
Continuing #290. 

**IMPORTANT**: I've not yet tested these changes, so while it's ready for a review, I'd like to run the e2e tests before merging.

### CODE REVIEW

**Comment by MindFlavor** [comment]
I love it ❤️.  Before going further, would you please try and use the just merged authorization policy? It should be enough to call `prepare_request_pipeline` and wrap the received context with `let mut pipeline_context = PipelineContext::new(ctx, ResourceType::Collections.into())`.

PS: I'll make sure to run the E2E tests on my account as soon as the integration tests will be green.

File: sdk/cosmos/src/clients/collection_client.rs [hunk 1]
```diff
@@ -60,8 +63,26 @@ impl CollectionClient {
...
+        ctx: Context,
+        document: &D,
+        options: CreateDocumentOptions<'_>,
+    ) -> Result<CreateDocumentResponse, crate::Error> {
+        let mut request = self
+            .cosmos_client()
+            .prepare_request_pipeline("collection", http::Method::POST);
+        let mut pipeline_context = PipelineContext::new(ctx, ResourceType::Collections.into());
+
+        options.decorate_request(&mut request, document)?;
```
Comment (MindFlavor) [inline sdk/cosmos/src/clients/collection_client.rs:77]: A couple of things to change:

1. According to [https://docs.microsoft.com/en-us/rest/api/cosmos-db/create-a-document](https://docs.microsoft.com/en-us/rest/api/cosmos-db/create-a-document), the URI should be `https://{databaseaccount}.documents.azure.com/dbs/{db-id}/colls/{coll-id}/docs`  (this is `https://{databaseaccount}.documents.azure.com/collection` instead).
2. The `ResourceType` is `Documents` and not `Collections` (if the resource type is wrong the authorization will fail).
3. The generated partition key is wrong. For example, running the example `document_00`, the expected header should be `"x-ms-documentdb-partitionkey": "[\"unique_id100\"]` but the generated one is `"x-ms-documentdb-partitionkey": "[{\"id\":\"unique_id100\",\"a_string\":\"Something here\",\"a_number\":100,\"a_timestamp\":1627459947}]"` which I believe is the whole document.

**Review by MindFlavor — CHANGES_REQUESTED** [review]
Nice work. 👍 
Unfortunately something is amiss because the generated requests aren't compliant with the specs (and as such they fail). 

I believe we should accept structs that implement both `CosmosEntity` and `Serialize` so we can easily extract the correct partition key.

**Comment by rylev** [comment]
@MindFlavor thanks for the review. I believe I've addressed all the issues you found. Can you review again? Thanks!

File: sdk/cosmos/src/clients/collection_client.rs [hunk 2]
```diff
@@ -137,4 +157,18 @@ impl CollectionClient {
...
         self.cosmos_client().http_client()
     }
+
+    pub(crate) fn pipeline(&self) -> &Pipeline<CosmosContext> {
+        self.cosmos_client().pipeline()
+    }
+
+    fn prepare_request_pipeline(&self, http_method: http::Method) -> Request {
+        let path = &format!(
+            "dbs/{}/colls/{}",
```
Comment (MindFlavor) [inline sdk/cosmos/src/clients/collection_client.rs:167]: This should be `"dbs/{}/colls/{}/docs"` as per [https://docs.microsoft.com/en-us/rest/api/cosmos-db/create-a-document](https://docs.microsoft.com/en-us/rest/api/cosmos-db/create-a-document)

**Review by MindFlavor — CHANGES_REQUESTED** [review]
Just a quick URI to fix and the tests pass! 

As a side note, the  `permission_token_usage` test fails but it's not due to your code. It's a regression introduced by the `AuthorizationPolicy` (or, better, by the fact that some methods use it and some don't). I will open a issue to tackle it in another PR.

**Review by MindFlavor — APPROVED** [review]
👍

**Pull request merged**